experimental/algorithm/LAGraph_RegularReachability.c

```diff
@@ -0,0 +1,329 @@
+    const GrB_Index *S,         // source vertices to start searching paths
+    size_t ns,                  // number of source vertices
+    char *msg
+)
+{
+
+    //--------------------------------------------------------------------------
+    // check inputs
+    //--------------------------------------------------------------------------
+
+    LG_CLEAR_MSG ;
+
+    GrB_Matrix frontier = NULL ;
+    GrB_Matrix symbol_frontier = NULL ;
+    GrB_Matrix next_frontier = NULL ;
+    GrB_Matrix visited = NULL ;
+    GrB_Vector final_reducer = NULL ;
+
+    GrB_Index ng = 0 ;                   // # nodes in the graph
+    GrB_Index nr = 0 ;                   // # states odes in the NFA
+    GrB_Index states = ns ;              // # pairs in the current
+                                         // correspondence beetwen the graph and
+                                         // the NFA
+
+    GrB_Index rows = 0 ;                 // utility matrix row count
+    GrB_Index cols = 0 ;                 // utility matrix column count
+    GrB_Index vals = 0 ;                 // utility matrix value count
+
+    GrB_Matrix *A = NULL ;
+    GrB_Matrix *AT = NULL ;
+    GrB_Matrix *B = NULL ;
+    GrB_Matrix *BT = NULL ;
+
+    LG_ASSERT (reachable != NULL && G != NULL && R != NULL && S != NULL,
+        GrB_NULL_POINTER) ;
+    (*reachable) = NULL ;
+
+    for (int32_t i = 0; i < nl; i++)
+    {
+        if (G[i] == NULL) continue ;
+        LG_TRY (LAGraph_CheckGraph (G[i], msg)) ;
+    }
+
+    for (int32_t i = 0; i < nl; i++)
+    {
+        if (R[i] == NULL) continue ;
+        LG_TRY (LAGraph_CheckGraph (R[i], msg)) ;
+    }
+
+    LG_TRY (LAGraph_Malloc ((void **) &A, nl, sizeof (GrB_Matrix), msg)) ;
+    LG_TRY (LAGraph_Malloc ((void **) &AT, nl, sizeof (GrB_Matrix), msg)) ;
+
+    for (int32_t i = 0; i < nl; i++) 
+    {
+        if (G[i] == NULL) 
+        {
+            A[i] = NULL ;
+            continue ;
+        }
+
+        A[i] = G[i]->A;
+        if (G[i]->kind == LAGraph_ADJACENCY_UNDIRECTED ||
+            G[i]->is_symmetric_structure == LAGraph_TRUE)
+        {
+            AT[i] = A[i] ;
+        }
+        else
+        {
+            AT[i] = G[i]->AT ;
+            LG_ASSERT_MSG (AT[i] != NULL, LAGRAPH_NOT_CACHED,
+                "G[i]->AT is required") ;
+        }
+    }
+
+    LG_TRY (LAGraph_Malloc ((void **) &B, nl, sizeof (GrB_Matrix), msg)) ;
+    LG_TRY (LAGraph_Malloc ((void **) &BT, nl, sizeof (GrB_Matrix), msg)) ;
+
+    for (int32_t i = 0; i < nl; i++)
+    {
+        if (R[i] == NULL)
+        {
+            B[i] = NULL ;
+            continue ;
+        }
+
+        B[i] = R[i]->A;
+        if (R[i]->kind == LAGraph_ADJACENCY_UNDIRECTED ||
+            R[i]->is_symmetric_structure == LAGraph_TRUE)
+        {
+            BT[i] = B[i] ;
+        }
+        else
+        {
+            BT[i] = R[i]->AT ;
+            LG_ASSERT_MSG (BT[i] != NULL, LAGRAPH_NOT_CACHED,
+                "R[i]->AT is required") ;
```

Comment:
It's not needed though it may speedup the execution. I've made it optional.

If it's `NULL` the descriptor is used instead.
```diff
@@ -212,9 +198,8 @@ int LAGraph_RegularPathQuery
         }
         else
         {
+            // BT[i] could be NULL and matrix will be transposed by the descriptor
             BT[i] = R[i]->AT ;
-            LG_ASSERT_MSG (BT[i] != NULL, LAGRAPH_NOT_CACHED,
-                "R[i]->AT is required") ;
         }
     }
@@ -338,8 +323,17 @@ int LAGraph_RegularPathQuery
             if (A[i] == NULL || B[i] == NULL) continue ;

             // Traverse the NFA
-            GRB_TRY (GrB_mxm (symbol_frontier, GrB_NULL, GrB_NULL,
-                GrB_LOR_LAND_SEMIRING_BOOL, BT[i], frontier, GrB_DESC_R)) ;
+            // Try to use a provided transposed matrix or use the descriptor
+            if (BT[i] != NULL)
+            {
+                GRB_TRY (GrB_mxm (symbol_frontier, GrB_NULL, GrB_NULL,
+                    GrB_LOR_LAND_SEMIRING_BOOL, BT[i], frontier, GrB_DESC_R)) ;
+            }
+            else
+            {
+                GRB_TRY (GrB_mxm (symbol_frontier, GrB_NULL, GrB_NULL,
+                    GrB_LOR_LAND_SEMIRING_BOOL, B[i], frontier, GrB_DESC_RT0)) ;
+            }
```

Reply:
Resolving, since discussed verbally Semyon Vyacheslavovich is ok.